test/test_distributed.py

```diff
@@ -700,13 +734,82 @@ def test_all_gather_multigpu(self):
+    def _test_DDP_helper(self, model, input_var, target, loss):
+        model.train()
+        output = model(input_var)
+        l = loss(output, target)
+        l.backward()
+
+    @unittest.skipIf(BACKEND != 'nccl' and BACKEND != 'gloo',
+                     "Only Nccl & Gloo backend support DistributedDataParallel")
+    @skip_if_no_cuda_distributed
+    @skip_if_no_multigpu
+    def test_DistributedDataParallel(self):
+        # Run a simple end to end DDP model, use result of single node model
+        # as baseline
+        group, group_id, rank = self._init_global_test()
+        rankToGPUMapping = self._init_multigpu_helper()
+
+        class Net(nn.Module):
+            def __init__(self):
+                super(Net, self).__init__()
+                self.features = nn.Linear(2, 4, bias=False)
+
+            def forward(self, x):
+                x = self.features(x)
+                return x
+
+        # cpu training setup
+        model = Net()
+        model_cpu = copy.deepcopy(model)
+
+        # single gpu training setup
+        model_gpu = copy.deepcopy(model)
+        gpu_subset = list(rankToGPUMapping[rank])
+        model_gpu.cuda(gpu_subset[0])
+
+        # DDP training setup
+        model_DDP = copy.deepcopy(model)
+        model_DDP.cuda(gpu_subset[0])
+        model_DDP = nn.parallel.DistributedDataParallel(model_DDP, device_ids=gpu_subset)
+
+        # batch_size for DDP should be divisible by #GPU per node.
+        batch_size = len(gpu_subset) * int(WORLD_SIZE)
+        input_cpu = torch.randn(batch_size, 1, 2)
+        target = torch.randn(batch_size, 1, 4)
+        loss = nn.MSELoss()
+
+        # cpu training
+        self._test_DDP_helper(model_cpu,
+                              Variable(input_cpu, requires_grad=True),
+                              Variable(target),
+                              loss)
+
+        # single gpu training
+        self._test_DDP_helper(model_gpu,
+                              Variable(input_cpu.cuda(gpu_subset[0]), requires_grad=True),
+                              Variable(target.cuda(gpu_subset[0], non_blocking=True)),
+                              loss)
+
+        # DDP training, DDP scatters subsets of input_cpu to nodes/GPUs
+        self._test_DDP_helper(model_DDP,
+                              Variable(input_cpu, requires_grad=True),
+                              Variable(target.cuda(gpu_subset[0], non_blocking=True)),
+                              loss)
+
+        self.assertEqual(model_cpu.features.weight.grad, model_gpu.features.weight.grad)
+        self.assertEqual(model_cpu.features.weight.grad, model_DDP.module.features.weight.grad)
+        self._barrier()
+
 if BACKEND == 'tcp' or BACKEND == 'gloo' or BACKEND == 'nccl':
     WORLD_SIZE = os.environ['WORLD_SIZE']
 
     class TestDistBackend(TestCase, _DistTestBase):
 
         MANAGER_PROCESS_RANK = -1
-        JOIN_TIMEOUT = 15
+        # DDP test on 8GPU machine takes about 21s
+        JOIN_TIMEOUT = 25
```

Comment:
Can you change this to a default of 15 that we used to have + a dict of optional per-test values that override the default?